pinot-core/src/main/java/org/apache/pinot/core/data/manager/realtime/SegmentBuildTimeLeaseExtender.java

```diff
@@ -46,39 +46,58 @@
   // Retransmit lease request 10% before lease expires.
   private static final int REPEAT_REQUEST_PERIOD_SEC = (EXTRA_TIME_SECONDS * 9 / 10);
   private static Logger LOGGER = LoggerFactory.getLogger(SegmentBuildTimeLeaseExtender.class);
-  private static final Map<String, SegmentBuildTimeLeaseExtender> INSTANCE_TO_LEASE_EXTENDER = new HashMap<>(1);
+  private static final Map<String, SegmentBuildTimeLeaseExtender> TABLE_TO_LEASE_EXTENDER = new HashMap<>();
+  private static ScheduledExecutorService _executor;
 
-  private ScheduledExecutorService _executor;
   private final Map<String, Future> _segmentToFutureMap = new ConcurrentHashMap<>();
   private final String _instanceId;
   private final ServerSegmentCompletionProtocolHandler _protocolHandler;
 
-  public static SegmentBuildTimeLeaseExtender getLeaseExtender(final String instanceId) {
-    return INSTANCE_TO_LEASE_EXTENDER.get(instanceId);
+  public static void initExecutor() {
+    _executor = new ScheduledThreadPoolExecutor(1);
+  }
+
+  public static void shutdownExecutor() {
+    if (_executor != null) {
+      _executor.shutdownNow();
+      _executor = null;
+    }
+  }
+
+  public static boolean isExecutorShutdown() {
+    return _executor == null;
   }
 
-  public static synchronized SegmentBuildTimeLeaseExtender create(final String instanceId,
+  public static synchronized SegmentBuildTimeLeaseExtender getLeaseExtender(final String tableNameWithType) {
+    return TABLE_TO_LEASE_EXTENDER.get(tableNameWithType);
+  }
+
+  public static synchronized SegmentBuildTimeLeaseExtender getOrCreate(final String instanceId,
```

Comment:
You may use `ConcurrentHashMap.compute()` which is atomic. Not very important though because it is not on query path